cluster-api/pkg/controller/machineset/controller.go

```diff
@@ -0,0 +1,218 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package machineset
+
+import (
+	"fmt"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	clusterv1alpha1 "k8s.io/kube-deploy/cluster-api/api/cluster/v1alpha1"
+	machineclientset "k8s.io/kube-deploy/cluster-api/client/clientset/versioned"
+	machinelistersv1alpha1 "k8s.io/kube-deploy/cluster-api/client/listers/cluster/v1alpha1"
+
+	"github.com/golang/glog"
+	"k8s.io/kube-deploy/cluster-api/api/cluster/v1alpha1"
+)
+
+// Controller is a implementation of a machine controller which creates machine resources
+type Controller struct {
+	// machineClient a client that knows how to consume Machine resources
+	machineClient machineclientset.Interface
+
+	// machineSetsInformer holds a shared informer for MachineSets
+	machineSetsInformer cache.SharedIndexInformer
+
+	// machineSetsLister holds a lister that knows how to list MachineSets from a cache
+	machineSetsLister machinelistersv1alpha1.MachineSetLister
+
+	// machineLister holds a lister that knows how to list Machines from a cache
+	machineLister machinelistersv1alpha1.MachineLister
+
+	// queue is where incoming work is placed to de-dup and to allow "easy"
+	// rate limited requeues on errors
+	queue workqueue.RateLimitingInterface
+}
+
+// New returns an instance of the MachineSet controller
+func New(machineClient machineclientset.Interface, machineSetsInformer cache.SharedIndexInformer, machineSetsLister machinelistersv1alpha1.MachineSetLister, machineLister machinelistersv1alpha1.MachineLister) *Controller {
```

Comment:
Please consider adding some tests. @karan started adding unit tests here, which could be used as reference: https://github.com/kubernetes/kube-deploy/pull/575

Reply:
Sure, I'll add a unit test.

Reply:
I've added some tests (check out `TestMachineSetControllerSyncHanlder`)